### DISCUSSION TITLE
Context propagation with Kafka Consumer

### DISCUSSION THREAD

**Discussion by mNantern** [open-telemetry/opentelemetry-java-instrumentation#4546] [Q&A] [answered]
Hello 👋.

I'm using the javaagent and everything is working fine except that I loose the trace after going trough Kafka.

The header `traceparent` is correctly inserted in Kafka but when the record is read by a consumer no trace_id is present in the MDC.

I see in [this discussion](https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/3860#discussioncomment-1207369) that calling `.get()` can result in issues and we are doing just that (to process the record per partition).

Do you know how I can manually inject the traceparent from the Kafka header?

🙏

**Comment by trask** [comment, marked as answer]
hi @mNantern!

You can read the `traceparent` from the Kafka message and create a scope based on it, and do your work under that scope, e.g.

```
String traceparent = ...

Context context =
    W3CTraceContextPropagator.getInstance()
        .extract(
            Context.root(),
            traceparent,
            new TextMapGetter<String>() {
              @Override
              public Iterable<String> keys(String carrier) {
                return Collections.singleton("traceparent");
              }
              @Override
              @Nullable
              public String get(String carrier, String key) {
                return key.equals("traceparent") ? carrier : null;
              }
            });

try (Scope scope = context.makeCurrent()) {
  // do your work here
}
```

**Reply by mNantern** [reply]
It's working perfectly, thank you 🙏
